libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -700,3 +705,88 @@ AP_GPS::_broadcast_gps_type(const char *type, uint8_t instance, int8_t baud_inde
     }
     GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, buffer);
 }
+
+
+/* 
+   re-assemble GPS_RTCM_DATA message
+ */
+void AP_GPS::handle_gps_rtcm_data(const mavlink_message_t *msg)
+{
+    mavlink_gps_rtcm_data_t packet;
+    mavlink_msg_gps_rtcm_data_decode(msg, &packet);
+
+    if (packet.len > MAVLINK_MSG_GPS_RTCM_DATA_FIELD_DATA_LEN) {
+        // invalid packet
+        return;
+    }
+    
+    if ((packet.flags & 1) == 0) {
+        // it is not fragmented, pass direct
+        inject_data_all(packet.data, packet.len);
+        return;
+    }
+
+    // see if we need to allocate re-assembly buffer
+    if (rtcm_buffer == nullptr) {
+        rtcm_buffer = (struct rtcm_buffer *)calloc(1, sizeof(*rtcm_buffer));
+        if (rtcm_buffer == nullptr) {
+            // nothing to do but discard the data
+            return;
+        }
+    }
+
+    uint8_t fragment = (packet.flags >> 1U) & 0x03;
+    uint8_t sequence = (packet.flags >> 3U) & 0x1F;
+
+    // see if this fragment is consistent with existing fragments
+    if (rtcm_buffer->fragments_received &&
+        (rtcm_buffer->sequence != sequence ||
```

Comment:
I agree that more buffers isn't a good solution. I think a better approach would be to only allow packets with a bigger sequence number. Then we have to address the part where it wraps around and my proposal to that is:
- if the new sequence number is smaller but is a maximum of 10 (?) numbers ahead (counting with the wrap around) then allow it and reset the buffer (this would be something like `32 - rtcm_buffer->sequence + sequence <= 10`)
- have a variable with the last time we inserted a fragment into the buffer and if more than a specified time has passed then reset the buffer and allow it.